tool/tsh/common/mcp.go

```diff
@@ -0,0 +1,238 @@
+		}
+
+		srv, err := newServerFunc(cf.Context, &dbmcp.NewServerConfig{
+			RootServer: server,
+			Databases:  databases,
+		})
+		if err != nil {
+			return trace.Wrap(err)
+		}
+		defer srv.Close(cf.Context)
+	}
+
+	return trace.Wrap(server.ServeStdio(cf.Context, cf.Stdin(), cf.Stdout()))
+}
+
+// closeLocalProxyFunc function used to close local proxy listeners.
+type closeLocalProxyFunc func() error
+
+// prepareDatabases based on the available MCP servers, initialize the database
+// local proxy and generate the MCP database.
+func (c *mcpDBCommand) prepareDatabases(
+	ctx context.Context,
+	registry dbmcp.Registry,
+	dbs []types.Database,
+	logger *slog.Logger,
+	tc *client.TeleportClient,
+	profile *client.ProfileStatus,
+	server *dbmcp.RootServer,
+) (map[string][]*dbmcp.Database, closeLocalProxyFunc, error) {
+	var (
+		dbsPerProtocol = make(map[string][]*dbmcp.Database)
+		closeFuncs     []closeLocalProxyFunc
+	)
+
+	for _, db := range dbs {
+		if !registry.IsSupported(db.GetProtocol()) {
+			logger.InfoContext(ctx, "database protocol unsupported, skipping it", "database", db.GetName(), "protocol", db.GetProtocol())
+			continue
+		}
+
+		route := tlsca.RouteToDatabase{
+			ServiceName: db.GetName(),
+			Protocol:    db.GetProtocol(),
+		}
+
+		listener, err := createLocalProxyListener("localhost:0", route, profile)
+		if err != nil {
+			logger.ErrorContext(ctx, "failed to start local proxy listener for database, skipping it", "database", db.GetName(), "error", err)
+			continue
+		}
+
+		cc := client.NewDBCertChecker(tc, route, nil, client.WithTTL(tc.KeyTTL))
+		lp, err := alpnproxy.NewLocalProxy(
+			makeBasicLocalProxyConfig(ctx, tc, listener, tc.InsecureSkipVerify),
+			alpnproxy.WithDatabaseProtocol(route.Protocol),
+			alpnproxy.WithMiddleware(cc),
+			alpnproxy.WithClusterCAsIfConnUpgrade(ctx, tc.RootClusterCACertPool),
+		)
+		if err != nil {
+			_ = listener.Close()
+			logger.ErrorContext(ctx, "failed to start local proxy for database, skipping it", "database", db.GetName(), "error", err)
+			continue
+		}
+		go func() {
+			defer lp.Close()
+			if err = lp.Start(ctx); err != nil {
+				logger.WarnContext(ctx, "failed to start local ALPN proxy", "error", err)
+			}
+		}()
+
+		mcpDB := &dbmcp.Database{
+			DB:           db,
+			DatabaseUser: c.databaseUser,
+			DatabaseName: c.databaseName,
+			Addr:         listener.Addr().String(),
+		}
+		dbsPerProtocol[db.GetProtocol()] = append(dbsPerProtocol[db.GetProtocol()], mcpDB)
+		server.RegisterDatabase(mcpDB)
+		closeFuncs = append(closeFuncs, listener.Close)
+	}
+
+	return dbsPerProtocol, func() error {
+		var errs []error
+		for _, closeFunc := range closeFuncs {
+			errs = append(errs, closeFunc())
+		}
+
+		return trace.NewAggregate(errs...)
+	}, nil
+}
+
+func (c *mcpDBCommand) getDatabases(ctx context.Context, sc *sharedDatabaseExecClient) ([]types.Database, error) {
+	labels, err := client.ParseLabelSpec(c.labels)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
```

Comment:
nit: can overwrite cf before making tc to avoid manually handling this? or, i would just pass `cf` to `newMCPDBCommand`.